File: iosApp/KaappiStudio/ViewModels/SettingsViewModel.swift
import SwiftUI
import shared

/// Main-actor front for the shared Kotlin `SettingsRepository`.
///
/// Storage keys, the theme fallback and the font-size range/clamping are the
/// repository's contract (`SettingsRepository.kt`); this class only mirrors
/// the values into published properties for SwiftUI.
@MainActor
class SettingsViewModel: ObservableObject {
    /// Font sizes the settings contract accepts, shared with Android.
    nonisolated static let fontSizeRange = Int(SettingsRepositoryKt.MIN_FONT_SIZE)...Int(SettingsRepositoryKt.MAX_FONT_SIZE)

    @Published var themeMode: ThemeMode {
        didSet { repository.setThemeMode(mode: themeMode) }
    }
    @Published var fontSize: Int {
        // The repository clamps into `fontSizeRange` before storing.
        didSet { repository.setFontSize(size: Int32(fontSize)) }
    }

    private let repository: SettingsRepository

    init(repository: SettingsRepository = SettingsRepository()) {
        self.repository = repository
        themeMode = repository.getThemeMode()
        fontSize = Int(repository.getFontSize())
    }

    var colorScheme: ColorScheme? {
        if themeMode == ThemeMode.light { return .light }
        if themeMode == ThemeMode.dark { return .dark }
        return nil
    }
}

File: app/src/main/java/com/kaappi/studio/viewmodel/SettingsViewModel.kt
package com.kaappi.studio.viewmodel

import androidx.lifecycle.ViewModel
import com.kaappi.studio.data.SettingsRepository
import com.kaappi.studio.domain.ThemeMode
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

class SettingsViewModel(private val settingsRepository: SettingsRepository) : ViewModel() {

    private val _themeMode = MutableStateFlow(settingsRepository.getThemeMode())
    val themeMode: StateFlow<ThemeMode> = _themeMode.asStateFlow()

    private val _fontSize = MutableStateFlow(settingsRepository.getFontSize())
    val fontSize: StateFlow<Int> = _fontSize.asStateFlow()

    fun setThemeMode(mode: ThemeMode) {
        settingsRepository.setThemeMode(mode)
        _themeMode.value = mode
    }

    fun setFontSize(size: Int) {
        settingsRepository.setFontSize(size)
        _fontSize.value = size
    }
}
